### ISSUE TITLE
[BUG] Async REST Timeout does not clear affected dialogs

### ISSUE THREAD

**Issue by spacetourist** [OpenSIPS/opensips#3239]
**OpenSIPS version you are running**

```
version: opensips 3.2.10 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, Q_MALLOC, F_MALLOC, HP_MALLOC, DBG_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: 4ef2e323b
main.c compiled on 00:00:00 May 27 2021 with gcc 11
```

**Describe the bug**
I am using async calls to rest_get as part of my new INVITE routing strategy however at times the remote web server is not responding in the allocated 5000ms. The dialog is initialised before this execution. There are many requests being processed at this time and each opensips instance is running ~25 HTTP requests/sec. What appears to be happening is that I see a few requests failing as expected with a return code of -3 which are then handled appropriately in the config script and the dialogs are removed. However several other requests appear to get orphaned any time this occurs - my logs show them issuing the HTTP request and then simply never returning, this leaves the dialogs stuck in memory until the instance is restarted.

The error I see for sessions which terminate normally is as expected:

`ERROR:rest_client:_resume_async_http_req: connected, but transfer timed out (5s)`

Those dialogs can be see to resume the script with return code -3 and then clear normally as I return a 504 upstream. This morning I have seen this twice on a single server however there are 6 or 7 dialogs that become stuck as the script never resumes. I detect these crudely using the CLI (note this brings back some dialogs which are in the process of establishing so it takes some effort to identify those which are stuck permanently, running it several times at intervals to extract those which never transition):

`opensips-cli -x mi dlg_list | jq '.Dialogs[] | select(.timeout==0 and .state==1 and .callee_cseq=="0") | .callid'`

The stuck dialogs present as follows:

```json
{
  "ID": "252.270ffe51",
  "db_id": "2551578620018",
  "state": 1,
  "user_flags": 0,
  "timestart": 0,
  "timeout": 0,
  "callid": "a78dff7a-73b1-4576-926e-7014baeff680",
  "from_uri": "sip:07500012345@192.168.163.107",
  "to_uri": "sip:6490227@sip.net",
  "caller_tag": "Uya271v7v02rN",
  "caller_contact": "sip:gw+rtc@192.168.163.107:5060;transport=udp;gw=rtc",
  "callee_cseq": "0",
  "caller_route_set": "",
  "caller_bind_addr": "udp:192.168.163.41:5060",
  "caller_sdp": "__THE_SDP__",
  "CALLEES": []
}
```

The stuck dialogs are few (although they add up, I have 1000 on one instance) and would probably have gone unnoticed however my config restricts concurrent calls by inspecting the other dialogs for the device via `get_dialogs_by_profile` so I am blocking transfers for users with stuck sessions.

**To Reproduce**
Example:
1. Start OpenSIPS with async rest_get in the new INVITE flow
2. Start traffic with a rate of 10 CPS
3. Introduce a sleep into the receiving web server
4. Notice that less `_resume_async_http_req` errors are produced than dialogs are affected, some never return

**Expected behavior**
All requests issued via async should time out individually and return -3, allowing the script to issue a 504 response and to tear down the dialog normally. 

If a fix is not feasible due to a weakness in curl_multi or similar then perhaps there should be some garbage collection that could clear these dialogs after a set period - these remain indefinitely with the current implementation.

**OS/environment information**
 - Operating System: AlmaLinux 9
 - OpenSIPS installation: Manually compiled at release tag 3.2.10

**Additional context**
I am not aware of any work that may have happened in this area since 3.2 however I am currently looking into an upgrade to 3.4 to see if this resolves the issue. The cause is clearly an external web service which is the immediate focus however I do believe that we have identified a weakness here which would be helpful to resolve.

**Comment by liviuchircu** [comment]
Hi, @spacetourist!  There have been some important fixes pushed to all of `3.2` and above versions regarding _rest_client_ timeouts, to be included in Wednesday's minor release:

* yes, it was possible for a `async(rest_get())` to hang indefinitely, if your `async()` statement did not include a timeout.  This was fixed, with the transfer timeout now defaulting to the `curl_timeout` modparam value now
* the lower of the `async()` statement timeout and `curl_timeout` will be correctly used to time out the transfer now

The fixes were done in two steps (1ea1852e + 1ecb3249), so make sure to grab them, re-test and let me know if the behavior improves.  Cheers!

**Comment by spacetourist** [comment]
Excellent, thanks for following up @liviuchircu !

I'll add that to the list and get it tested as soon as feasible. All the best :v:
